Roxygen integration (wrapper generation using roclets)

Ilia-Kosenkov opened the pull request:
This continues the discussions from #56, #59, #63, #64.
In this PR we make use of roxygen2 roclet feature, effectively defining a custom script that is injected in the middle of `roxygen2::roxygenize()` process.

Primary modifications:
1. Improved `use_extendr` to write additional metadata necessary to enable roclets support
2. Implemented `registration_roclet` which simply calls `rextendr::register_extendr()`.
3. Implemented a roclet to track changes in Rust files, similar to what `pkgbuild::needs_compilation()` does.

Ilia-Kosenkov commented:
Documentation is not properly updated so the main CI fails, I will fix it.

Ilia-Kosenkov commented:
It is hard to showcase how file tracking is working.
Imagine that I setup an empty rextendr project, compile it, then modify (touch) `lib.rs` and `Cargo.toml`.
Here is what happens at the next `devtools::document()`:
```
Updating testpkg documentation
i Loading testpkg
Generating extendr wrapper functions for package: testpkg
Writting wrappers to:
C:/Users/[redacted]/source/repos/testpkg/R/extendr-wrappers.R
v Writting wrappers to R/extendr-wrappers.R.
i File C:/Users/[redacted]/source/repos/testpkg/src/rust/Cargo.toml has been modified since last compilation.
i File C:/Users/[redacted]/source/repos/testpkg/src/rust/src/lib.rs has been modified since last compilation.
! Run `devtools::document()` one more time to recompile Rust source.

```

clauswilke commented on R/filetracker_roclet.R:
```diff
@@ -0,0 +1,22 @@
+  roxygen2::roclet("filetracker")
+}
+
+#' @method roclet_process roclet_filetracker
```
I believe `@method` is deprecated. Do you know whether it is needed here?

Ilia-Kosenkov replied:
I had problems with exporting this S3 method. The generic itself comes from roxygen2, and I need an `S3method` entry in the NAMESPACE of rextendr. Without it, generics are not resolved correctly.

Ilia-Kosenkov replied:
UPD: I found this topic here which deals with exactly the same problem:
https://github.com/r-lib/devtools/issues/1585
`@method` is not required as long as `document()` is executed twice. I believe this is the same problem that we have -- at the time of first compilation, there is no knowledge that `S3` method from `roxygen2` is imported and exported methods should be treated as such. During the second `document()` call, it is already known and the correct `S3` export is generated.

I do not know why it is still a thing, it looks like it has been fixed and the issue is several years old, yet here we are.

clauswilke commented on R/track_rust_source.R:
```diff
@@ -0,0 +1,80 @@
+}
+
+get_rust_files <- function(path = ".") {
+  rust_root <- rprojroot::find_package_root_file("src", "rust", path = path)
```
I would either not get into this at all (it'll never be 100% right, since we're not tracking dependencies in the way `cargo` does) or recursively find all `*.rs` and `Cargo.toml` files under `src`.

Ilia-Kosenkov replied:
We of course cannot outdo `cargo` or `make` itself. What we can do is track 'obvious' changes, like modifying rust source or `Cargo.toml`. I will remove `rust` subdir from path -- any `*.rs` file under `pkg_root/src/` should be tracked.

clauswilke commented on R/track_rust_source.R:
```diff
@@ -0,0 +1,80 @@
+needs_compilation <- function(path = ".") {
+  library_path <- get_library_path(path)
+
+  # This will likely never happen
```
Why not?

Ilia-Kosenkov replied:
I originally thought it may hit this. I think that `roxygen2::roxygenize()` invokes pkgbuild, and `pkgbuild::needs_compilation()` checks if the library file is present, and triggers recompilation if it is absent. This happens **before** any roclets are exectued, so even if there is no library before `roxygen2::roxygenize()`, by the time it invokes our roclet, the library has been already built.
The (unlikely) scenario is the library gets removed during `roxygenize()` process. If that happens, I would not trust the output of the pipeline at all.

clauswilke commented on R/filetracker_roclet.R:
```diff
@@ -0,0 +1,22 @@
+roclet_output.roclet_filetracker <- function(x, results, base_path, ...) {
+  if (needs_compilation()) {
+    touch_makevars()
+    cli::cli_alert_warning("Run {.code devtools::document()} one more time to recompile Rust source.")
```
Why run `document()` again and not `install()`? Wouldn't that be the normal sequence, or am I missing something?

Ilia-Kosenkov replied:
My understanding is that when we change `*.rs`, which is not tracked by the default build system, the roclet discovers something has been modified at the end of the pipeline. We `touch` `Makevars`, but we have no way to ask for recompilation, as this basically requires the whole pipeline to be rerun. 

In short, all roclets are executed **after** compilation has happenned and their execution is independent from each other. 
If we update wrappers in roclet, the wrapers and their tags will not be read by the roxygen2 roclet which emits documentation and `NAMESPACE` entries, and there is no way to bypass this without hijacking the build system. The same applies to the source code modification.

clauswilke commented on R/registration_roclet.R:
```diff
@@ -0,0 +1,21 @@
+#' Create roclet that registers `extendr` package
```
"Create *a* roclet that registers *the* `extendr` package"

clauswilke commented on R/use_roclets.R:
```diff
@@ -0,0 +1,63 @@
+
+#' Use roclets to augment package compilaion.
```
spelling: compilation

clauswilke commented on R/use_roclets.R:
```diff
@@ -0,0 +1,63 @@
+#' Use roclets to augment package compilaion.
+#'
+#' @param use_roxygen_roclets Logical (default: `TRUE`), indicates
+#' if default `roxygen2` roclets should be referenced in the DESCRIPTION file.
```
Not sure I understand what this means.

Ilia-Kosenkov replied:
roxygen2 provides several `default` roclets, e.g. `roxygen2::rd_roclet` and `roxygen2::namespace_roclet`. These are responsible for parsing tags and writing docs & exports. If we specify custom roclets in `DESCRIPTION`, these default ones are not called.
[Here](https://github.com/extendr/rextendr/issues/59#issuecomment-796728016) I got an explanation of how to add these default roclets:
> 
> 
> @Ilia-Kosenkov you can get the default roclets with `roxygen2::roxy_meta_get("roclets")`, so something like `Roxygen: list(roclets = c(rextendr::my_roclet, roxygen2::roxy_meta_get("roclets")))` should run the rextendr roclet along with the default roclets.

So, this flag controls if `roxygen2::roxy_meta_get("roclets")` is written to the `DESCRIPTION` file. Not sure we will ever need this, but it is good to have configurable options.

Anyway, I'll write more comments & explanations.

clauswilke reviewed (commented):
I left various comments throughout. Overall, there's a lot of new code here that interacts in complex ways, so it's a bit hard to figure out how it all connects. I think it would be important to add more comments to explain how the pieces fit together. For example, all the various helper functions should have some brief documentation describing what they do and how they are used.

Ilia-Kosenkov commented:
@clauswilke,
Sure, it was just a quick one-evening prototype. I will update it.

clauswilke commented:
@Ilia-Kosenkov @yutannihilation Is there any sort of interaction between #64 and #65 or do they solve distinct problems and can be merged separately? Is there a preferred order in which they should be merged?

Ilia-Kosenkov commented:
@clauswilke,
I believe this was a reaction from @yutannihilation to my comment that there is an issue with library unloading ([quoting you](https://github.com/extendr/rextendr/issues/59#issuecomment-796298760)):
> 
> 
> > The library can always be loaded/unloaded just for the sake of the call to make_wrappers.
> 
> I have not been able to do so reliably on macOS. It's possible somebody else can figure this out. But, the documentation of `dyn.unload()` contains this sentence: "Note that unloading a DLL and then re-loading a DLL of the same name may or may not work"

I personally have not experienced that so I have no solution to the problem I do not understand.
With that said, #64 is independent from this PR. This PR does not touch `register_extendr`, it only injects a call to this method into roxygen2 pipeline.

Ilia-Kosenkov commented on R/track_rust_source.R:
```diff
@@ -0,0 +1,80 @@
+# See https://github.com/r-lib/pkgbuild/blob/2aaecf838eb023b788b7a31e267ac644d616dae0/R/compile-dll.R
```
It is a reminder (mostly for myself). It points to the pkgbuild implementation, from which I drew inspiration. I did not copy any code though. pkgbuild is MIT.
This can be removed.

Ilia-Kosenkov commented:
<details>
<summary>Sample workflow</summary>

``` r
tempdir <- tempdir()
pkg_root <- fs::path(tempdir, "samplepkg")
fs::dir_create(pkg_root, recurse = TRUE)
setwd(pkg_root)
devtools::create(".")
#> v Setting active project to 'C:/Users/[redacted]/AppData/Local/Temp/RtmpobYxgU/samplepkg'
#> v Creating 'R/'
#> v Writing 'DESCRIPTION'
#> Package: samplepkg
#> Title: What the Package Does (One Line, Title Case)
#> Version: 0.0.0.9000
#> Authors@R (parsed):
#>     * First Last <first.last@example.com> [aut, cre] (YOUR-ORCID-ID)
#> Description: What the package does (one paragraph).
#> License: `use_mit_license()`, `use_gpl3_license()` or friends to pick a
#>     license
#> Encoding: UTF-8
#> LazyData: true
#> Roxygen: list(markdown = TRUE)
#> RoxygenNote: 7.1.1
#> v Writing 'NAMESPACE'
#> v Setting active project to '<no active project>'
usethis::proj_activate(".") # Bypasses an issue with non-interactive session
#> v Setting active project to 'C:/Users/[redacted]/AppData/Local/Temp/RtmpobYxgU/samplepkg'
rextendr::use_extendr(use_roclets = TRUE)
#> Creating `src` ...
#> Creating `R/extendr-wrappers.R` ...
#> v Adding 'useDynLib(samplepkg, .registration = TRUE)' to 'NAMESPACE'
#> v Adding roclets to `Roxygen` field in DESCRIPTION.
#> Done.
#> 
#> Please run `devtools::document()` for changes to take effect.
#> Also update the system requirements in your `DESCRIPTION` file.
devtools::document()
#> Updating samplepkg documentation
#> i Loading samplepkg
#> Re-compiling samplepkg
#>   -  installing *source* package 'samplepkg' ... (575ms)
#>      ** using staged installation
#>      ** libs
#>      rm -Rf samplepkg.dll ./rust/target/x86_64-pc-windows-gnu/release/libsamplepkg.a entrypoint.o
#>      "C:/rtools40/mingw64/bin/"gcc  -I"C:/PROGRA~1/R/R-40~1.4/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c entrypoint.c -o entrypoint.o
#>      cargo build --target=x86_64-pc-windows-gnu --lib --release --manifest-path=./rust/Cargo.toml
#>              Updating crates.io index
#>             Compiling winapi-x86_64-pc-windows-gnu v0.4.0
#>       Compiling winapi-build v0.1.1
#>       Compiling winapi v0.3.9
#>       Compiling winapi v0.2.8
#>             Compiling proc-macro2 v1.0.24
#>             Compiling unicode-xid v0.2.1
#>             Compiling syn v1.0.63
#>             Compiling extendr-engine v0.2.0
#>             Compiling lazy_static v1.4.0
#>             Compiling kernel32-sys v0.2.2
#>             Compiling quote v1.0.9
#>             Compiling extendr-macros v0.2.0
#>             Compiling libR-sys v0.2.1
#>             Compiling extendr-api v0.2.0
#>             Compiling samplepkg v0.1.0 (C:\Users\[redacted]\AppData\Local\Temp\RtmpobYxgU\samplepkg\src\rust)
#>              Finished release [optimized] target(s) in 1m 09s
#>      C:/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o samplepkg.dll tmp.def entrypoint.o -L./rust/target/x86_64-pc-windows-gnu/release -lsamplepkg -lws2_32 -ladvapi32 -luserenv -LC:/PROGRA~1/R/R-40~1.4/bin/x64 -lR
#>      installing to C:/Users/[redacted]/AppData/Local/Temp/RtmpobYxgU/devtools_install_53bc7ff64e8c/00LOCK-samplepkg/00new/samplepkg/libs/x64
#>   -  DONE (samplepkg)
#> 
#> i Generating extendr wrapper functions for package: samplepkg.
#> v Writting wrappers to R/extendr-wrappers.R.
#> ! Run `devtools::document()` one more time to update documentation and NAMESPACE.
devtools::document()
#> Updating samplepkg documentation
#> i Loading samplepkg
hello_world()
#> [1] "Hello world!"
brio::write_lines(
    c(
        "use extendr_api::prelude::*;",
        "/// Generated on the fly",
        "/// @returns A string",
        "/// @export",
        "#[extendr]",
        "fn hello_world() -> &'static str {",
        "   \"Hello world! Modified!\"",
        "}",
        "extendr_module! {",
        "   mod samplepkg;",
        "   fn hello_world;",
        "}"
    ),
    path = fs::path(pkg_root, "src", "rust", "src", "lib.rs")
)
fs::file_touch(fs::path(pkg_root, "src", "rust", "Cargo.toml"))
devtools::document() # Detects changes to Rust
#> Updating samplepkg documentation
#> i Loading samplepkg
#> i File src/rust/Cargo.toml has been modified since last compilation.
#> i File src/rust/src/lib.rs has been modified since last compilation.
#> ! Run `devtools::document()` one more time to recompile Rust source.
devtools::document() # Recompiles & updates wrappers
#> Updating samplepkg documentation
#> i Loading samplepkg
#> Re-compiling samplepkg
#>   -  installing *source* package 'samplepkg' ... (547ms)
#>      ** using staged installation
#>      ** libs
#>      rm -Rf samplepkg.dll ./rust/target/x86_64-pc-windows-gnu/release/libsamplepkg.a entrypoint.o
#>      "C:/rtools40/mingw64/bin/"gcc  -I"C:/PROGRA~1/R/R-40~1.4/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c entrypoint.c -o entrypoint.o
#>      cargo build --target=x86_64-pc-windows-gnu --lib --release --manifest-path=./rust/Cargo.toml
#>             Compiling samplepkg v0.1.0 (C:\Users\[redacted]\AppData\Local\Temp\RtmpobYxgU\samplepkg\src\rust)
#>              Finished release [optimized] target(s) in 1.31s
#>      C:/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o samplepkg.dll tmp.def entrypoint.o -L./rust/target/x86_64-pc-windows-gnu/release -lsamplepkg -lws2_32 -ladvapi32 -luserenv -LC:/PROGRA~1/R/R-40~1.4/bin/x64 -lR
#>      installing to C:/Users/[redacted]/AppData/Local/Temp/RtmpobYxgU/devtools_install_53bc79197f8f/00LOCK-samplepkg/00new/samplepkg/libs/x64
#>   -  DONE (samplepkg) (355ms)
#> 
#> i Generating extendr wrapper functions for package: samplepkg.
#> v Writting wrappers to R/extendr-wrappers.R.
#> ! Run `devtools::document()` one more time to update documentation and NAMESPACE.
devtools::document() # Updates exports & NAMESPACE using new wrappers
#> Updating samplepkg documentation
#> i Loading samplepkg
hello_world()
#> [1] "Hello world! Modified!"
```

<sup>Created on 2021-03-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

</details>

yutannihilation commented:
> Is there a preferred order in which they should be merged?

I have no words to add to @Ilia-Kosenkov's comment, thanks for summarizing.

yutannihilation reviewed (commented):
Hmm, I really appreciate your effort here, but is this really worth this complexity...? (Sorry, I'm not sure if I have a right to complain here after proposing another complexity on #63 :roll_eyes:)

I especially feel this part of your sample workflow frustrating in that we need to `devtools::document()` three times. Do we really need to get everything done in `devtools::document()`? Although it 's a non-standard command, I guess we can implement `rextendr::document()`, which covers all the steps in a single call.

``` r
devtools::document() # Detects changes to Rust
#> Updating samplepkg documentation
#> i Loading samplepkg
#> i File src/rust/Cargo.toml has been modified since last compilation.
#> i File src/rust/src/lib.rs has been modified since last compilation.
#> ! Run `devtools::document()` one more time to recompile Rust source.
devtools::document() # Recompiles & updates wrappers
#> Updating samplepkg documentation
#> i Loading samplepkg
#> Re-compiling samplepkg
#>   -  installing *source* package 'samplepkg' ... (547ms)
#>      ** using staged installation
#>      ** libs
#>      rm -Rf samplepkg.dll ./rust/target/x86_64-pc-windows-gnu/release/libsamplepkg.a entrypoint.o
#>      "C:/rtools40/mingw64/bin/"gcc  -I"C:/PROGRA~1/R/R-40~1.4/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c entrypoint.c -o entrypoint.o
#>      cargo build --target=x86_64-pc-windows-gnu --lib --release --manifest-path=./rust/Cargo.toml
#>             Compiling samplepkg v0.1.0 (C:\Users\[redacted]\AppData\Local\Temp\RtmpobYxgU\samplepkg\src\rust)
#>              Finished release [optimized] target(s) in 1.31s
#>      C:/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o samplepkg.dll tmp.def entrypoint.o -L./rust/target/x86_64-pc-windows-gnu/release -lsamplepkg -lws2_32 -ladvapi32 -luserenv -LC:/PROGRA~1/R/R-40~1.4/bin/x64 -lR
#>      installing to C:/Users/[redacted]/AppData/Local/Temp/RtmpobYxgU/devtools_install_53bc79197f8f/00LOCK-samplepkg/00new/samplepkg/libs/x64
#>   -  DONE (samplepkg) (355ms)
#> 
#> i Generating extendr wrapper functions for package: samplepkg.
#> v Writting wrappers to R/extendr-wrappers.R.
#> ! Run `devtools::document()` one more time to update documentation and NAMESPACE.
devtools::document() # Updates exports & NAMESPACE using new wrappers
#> Updating samplepkg documentation
#> i Loading samplepkg
```

Ilia-Kosenkov commented on .github/workflows/test_pkg_gen.yaml:
```diff
@@ -3,6 +3,7 @@ on:
     branches:
       - main
       - master
+      - '*'
```
Sure, I'll remove it. I was not expecting this will make it into PR draft.

clauswilke commented:
> Hmm, I really appreciate your effort here, but is this really worth this complexity...? 

I have to say I feel the same. A lot of complexity and technical debt, not just for us but also our endusers (e.g., roclet statements in the `DESCRIPTION`; if devtools changes their default roclets, users will have to manually update their `DESCRIPTION` or more likely won't do it and will get strange problems). I also feel there's a lot of code that works around current limitations in the extension mechanisms of devtools, and I don't like that strategy. I'd rather work with the devtools developers directly to modify their code so it supports rextendr natively (or create better hooks). That is certainly going to be an option if the extendr project takes off and gains a broad userbase.

As is, I'd rather limit ourselves to `rextendr::register_extendr()`, which does what it currently does, and `rextendr::document()`, which will check if compilation is needed, compile, run `rextendr::register_extendr()`, and then run `devtools::document(). That's a simple and straightforward setup that users can grasp easily and it's good enough for where the project currently stands.

Ilia-Kosenkov commented:
@clauswilke, @yutannihilation, 
I have no objections. I was interested if this can be achieved at all (with the current limitations). This PR can be closed. 

It would be nice to have at least {pkgbuild} support for tracking rust sources. This is an important first step, without which wrapper generation becomes unreliable. Perhaps they will consider accepting a PR at some point?

clauswilke commented:
@Ilia-Kosenkov Sounds good. Why don't you take the Rust source file tracking code you wrote here and implement a version of `rextendr::document()`?

The end product should be something like this, I believe (just a rough draft for illustration):
```
document <- function() {
  if (rextendr::needs_compilation()) {
    pkgbuild::compile_dll()
  }
  rextendr::register_extendr()
  devtools::document()
}
```

Maybe I'm forgetting something.

Ilia-Kosenkov commented:
@clauswilke,
Yeah, I was thinking about something like this. 
I will make a separate PR based on this code and prototype there.

clauswilke commented:
@Ilia-Kosenkov You may want to add something like the following if the rstudio api is available:
https://github.com/r-lib/devtools/blob/d029f66552569a7842fa39ebde089394d86c9e06/R/save-all.R#L9-L13

Unfortunately devtools doesn't export this, but it's only two lines anyways, so doesn't matter much.

clauswilke commented:
Devtools calls this before doing anything:
https://github.com/r-lib/devtools/blob/d029f66552569a7842fa39ebde089394d86c9e06/R/document.R#L19

Ilia-Kosenkov commented:
@clauswilke,
Should we have an explicit dependence on {rstudioapi}, or perhaps we can use something like `requireNamespace`?

clauswilke commented:
I think it should be an optional dependency on rstudioapi, i.e., put it into suggests not imports.

clauswilke commented:
Can we close this PR?

Ilia-Kosenkov commented:
Closing in favor of #67 .
